File: Ahorn/entities/gravityRefill.jl
# Copyright (c) Shane Woolcock. Licensed under the MIT Licence.
# See the LICENCE file in the repository root for full licence text.

module GravityHelperGravityRefill

using ..Ahorn, Maple

const PLUGIN_VERSION = "1"

@mapdef Entity "GravityHelper/GravityRefill" GravityRefill(
    x::Integer, y::Integer,
    pluginVersion::String=PLUGIN_VERSION,
    charges::Integer=1,
    dashes::Integer=-1,
    oneUse::Bool=false,
    refillsDash::Bool=true,
    refillsStamina::Bool=true,
    respawnTime::Float64=2.5
)

const placements = Ahorn.PlacementDict(
    "Gravity Refill (GravityHelper)" => Ahorn.EntityPlacement(
        GravityRefill
    ),
    "Gravity Refill (Single Use) (GravityHelper)" => Ahorn.EntityPlacement(
        GravityRefill,
        "point",
        Dict{String, Any}(
            "oneUse" => true
        )
    ),
    "Gravity Refill (No Dash/Stamina) (GravityHelper)" => Ahorn.EntityPlacement(
        GravityRefill,
        "point",
        Dict{String, Any}(
            "refillsDash" => false,
            "refillsStamina" => false
        )
    ),
    "Gravity Refill (No Dash/Stamina, Single Use) (GravityHelper)" => Ahorn.EntityPlacement(
        GravityRefill,
        "point",
        Dict{String, Any}(
            "refillsDash" => false,
            "refillsStamina" => false,
            "oneUse" => true
        )
    ),
    "Gravity Refill (Two Dashes/Charges) (GravityHelper)" => Ahorn.EntityPlacement(
        GravityRefill,
        "point",
        Dict{String, Any}(
            "dashes" => 2,
            "charges" => 2
        )
    ),
    "Gravity Refill (Two Dashes/Charges, Single Use) (GravityHelper)" => Ahorn.EntityPlacement(
        GravityRefill,
        "point",
        Dict{String, Any}(
            "dashes" => 2,
            "charges" => 2,
            "oneUse" => true
        )
    ),
)

const normalSprite = "objects/GravityHelper/gravityRefill/idle00"
const noDashSprite = "objects/GravityHelper/gravityRefill/idle_no_dash00"
const twoDashSprite = "objects/GravityHelper/gravityRefill/idle_two_dash00"

function spriteForEntity(entity::GravityRefill)
    local refillsDash = get(entity.data, "refillsDash", true)
    local dashes = get(entity.data, "dashes", -1)
    return refillsDash && dashes == 2 ? twoDashSprite : !refillsDash ? noDashSprite : normalSprite
end

Ahorn.editingIgnored(entity::GravityRefill, multiple::Bool=false) = multiple ? String["x", "y", "modVersion", "pluginVersion"] : String["modVersion", "pluginVersion"]

function Ahorn.selection(entity::GravityRefill)
    x, y = Ahorn.position(entity)
    local sprite = spriteForEntity(entity)
    return sprite == twoDashSprite ? Ahorn.Rectangle(x - 5, y - 7, 10, 14) : Ahorn.Rectangle(x - 4, y - 5, 8, 10)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::GravityRefill, room::Maple.Room)
    local sprite = spriteForEntity(entity)
    Ahorn.drawSprite(ctx, sprite, 0, 0)
end

end

File: Loenn/entities/gravityRefill.lua
-- Copyright (c) Shane Woolcock. Licensed under the MIT Licence.
-- See the LICENCE file in the repository root for full licence text.

local utils = require("utils")
local consts = require("mods").requireFromPlugin("consts")
local helpers = require("mods").requireFromPlugin("helpers")

local placementData = helpers.createPlacementData('2', {
    charges = 1,
    dashes = -1,
    oneUse = false,
    refillsDash = true,
    refillsStamina = true,
    respawnTime = 2.5,
    legacyRefillBehavior = false,
})

local gravityRefill = {
    name = "GravityHelper/GravityRefill",
    depth = -100,
    ignoredFields = consts.ignoredFields,
    fieldInformation = {
        charges = {
            fieldType = "integer",
        },
        dashes = {
            fieldType = "integer",
        },
        respawnTime = {
            fieldType = "number",
        },
    },
    fieldOrder = {
        "x", "y",
        "charges", "respawnTime",
        "dashes", "refillsDash", "refillsStamina",
        "oneUse", "legacyRefillBehavior",
    },
    placements = {
        {
            name = "normal",
            ignoredFields = consts.ignoredFields,
            data = helpers.union(placementData),
        },
        {
            name = "normalSingleUse",
            ignoredFields = consts.ignoredFields,
            data = helpers.union(placementData, {
                oneUse = true,
            }),
        },
        {
            name = "twoDash",
            ignoredFields = consts.ignoredFields,
            data = helpers.union(placementData, {
                dashes = 2,
                charges = 2,
            }),
        },
        {
            name = "twoDashSingleUse",
            ignoredFields = consts.ignoredFields,
            data = helpers.union(placementData, {
                dashes = 2,
                charges = 2,
                oneUse = true,
            }),
        },
        {
            name = "noDash",
            ignoredFields = consts.ignoredFields,
            data = helpers.union(placementData, {
                refillsDash = false,
                refillsStamina = false,
            }),
        },
        {
            name = "noDashSingleUse",
            ignoredFields = consts.ignoredFields,
            data = helpers.union(placementData, {
                refillsDash = false,
                refillsStamina = false,
                oneUse = true,
            }),
        },
    },
}

function gravityRefill.texture(room, entity)
    local suffix = ""
    if entity.refillsDash and entity.dashes == 2 then
        suffix = "_two_dash"
    elseif not entity.refillsDash then
        suffix = "_no_dash"
    end

    return "objects/GravityHelper/gravityRefill/idle"..suffix.."00"
end

function gravityRefill.selection(room, entity)
    return utils.rectangle(entity.x - 4, entity.y - 5, 8, 10)
end

return gravityRefill
